### ISSUE TITLE
Need more clarification/derivation on the “2”in 2*pi*(1-pi) Hessian of MultiClass:Softmax 

### ISSUE THREAD

**Issue by hrbzkm98** [dmlc/xgboost#5299]
Here is a list of issues raised in the past: #638, #1825, #2024, #2485
Here is a paper on where the 2 comes from: http://homes.cs.washington.edu/~tqchen/data/pdf/GBCRF-AISTATS15.pdf

However the paper is on CRF, and after reading am still confused why the Hessian is 2 * pi(1-pi). 
My understanding is that, as long as we have a “majorizing” hessian, we are guaranteed to converge when optimizing quadratic approximation. My guess is that, since Hessian is diagonal, with positive entries (hence positive semi-definite), we could multiply the original hessian by any number greater than one, hence “2” is a choice; or like that in Poisson regression #2661 exp(p+max_delta_step). Is that the case? What is the acceptable range for this constant and how do we choose this constant? Since pi(1-pi) is bounded above by 0.25, can we just set the Hessian to be 0.25 regardless of pi? How should we handle the case where hessian is not positive definite?

The principal reason is we do not want Hessian to be zero. However, when pi*(1-pi) is very small, e.g. 10e-10, a factor of 2 doesn’t seem to help.

I am asking this question because I am implementing my own custom objective function and I am very confused on the choice of constant to multiply by my hessian matrix. This could serve as a future guideline for implementing custom hessian. It would be very helpful to have some kind of formal derivation but intuitive explanation helps too.

**Comment by trivialfis** [comment]
I asked a similar question before and have been wanting to write a small doc about it.  Let me see if I can do it on spare time this week.

**Comment by hrbzkm98** [comment]
Thank you so much, @trivialfis, please kindly let me know when you have something ready. I am eager to know the answer.

**Comment by trivialfis** [comment]
@hrbzkm98 There is a strong relationship between logistic model and conditional random field.  Quoting from #1815 , softmax is a special case of CRF with no edge potential, you can remove all edge potential in all equations in the CRF paper you referenced.  It will lead to the hessian defined in XGBoost.
